### PR TITLE
Filter out only link tags with rel attributes that cause browser activity

### LINKED ISSUE #305: BTR should be more selective about which `link` tags to ignore from head
**Bug**

I am currently unable to add a `<link>` element to the `<head>` contents. This is because the btr process currently strips out all `<link>` tags that are not present in the initial HTML.
We should instead consider allowing `<link>` tags with `rel` attributes set to those which do not kick off browser fetches etc. The `rel` attributes to filter out are
- 'preconnect'
- 'prefetch'
- 'preload'
- 'prerender'
- 'dns-prefetch'
- 'stylesheet'

### PR DESCRIPTION
**Pull request by tomdye**
**Type:** feature

The following has been addressed in the PR:

* [x] There is a related issue
* [x] All code has been formatted with [`prettier`](https://prettier.io/) as per the [readme code style guidelines](./../#code-style)
* [x] Unit or Functional tests are included in the PR

**Description:**

Filters out links with rel attributes which cause browser activity, ie. fetches etc.

Resolves #305

### CODE REVIEW

File: src/build-time-render/BuildTimeRender.ts [hunk 1]
```diff
@@ -113,6 +113,18 @@ class MockCompilation {
 	}
 }
 
+const relRegExp = /rel\=\"([\w]+)\"/;
```
Comment (agubler) [inline src/build-time-render/BuildTimeRender.ts:116]: Could this be changed to be slightly simpler?

```ts
const.dynamicLinkRegExp = /rel\=(\"|\')(preconnect|prefetch|preload|prerender|dns-prefetch|stylesheet)(\"|\')/;
```

and then used like `dynamicLinkRegExp.test(linkString)` in the places? 

Also as it stands the regex actually wouldn't work for `dns-prefetch`

**Comment by tomdye** [comment]
Comment addressed @agubler , thank you for that

File: src/build-time-render/BuildTimeRender.ts [hunk 2]
```diff
@@ -351,6 +353,12 @@ export default class BuildTimeRender {
 		let pathValue = typeof path === 'object' ? path.path : path;
 		let content = await getForSelector(page, `#${this._root}`);
 		let head = await getAllForSelector(page, 'head > *:not(script):not(link)');
+
+		let links = await getAllForSelector(page, 'head > link');
+		links = links.filter((link) => {
```
Comment (agubler) [inline src/build-time-render/BuildTimeRender.ts:358]: inline the filter?

```ts
links = links.filter((link) => !dynamicLinkRegExp.test(link);
```

File: src/build-time-render/BuildTimeRender.ts [hunk 3]
```diff
@@ -643,9 +651,16 @@ ${blockCacheEntry}`
 		const root = parse(html, { script: true, style: true });
 		const rootNode = root.querySelector(`#${this._root}`);
 		const headNode = root.querySelector('head');
+
 		if (headNode) {
 			this._headNodes = (headNode.childNodes.filter((node) => node.nodeType === 1) as HTMLElement[])
-				.filter((node) => node.tagName === 'script' || node.tagName === 'link')
+				.filter((node) => {
```
Comment (agubler) [inline src/build-time-render/BuildTimeRender.ts:657]:
```ts
filter((node) => node.tagName === 'script' || (node.tagName === 'link' && dynamicLinkRegExp.test(node.toString()))
```

**Review by agubler — APPROVED** [review]

**Pull request merged**